ext/standard/array.c

```diff
@@ -3694,13 +3694,11 @@ PHPAPI int php_array_merge_recursive(HashTable *dest, HashTable *src) /* {{{ */
 				ZVAL_DEREF(src_zval);
 				ZVAL_DEREF(dest_zval);
 				thash = Z_TYPE_P(dest_zval) == IS_ARRAY ? Z_ARRVAL_P(dest_zval) : NULL;
-				if ((thash && GC_IS_RECURSIVE(thash)) || (src_entry == dest_entry && Z_ISREF_P(dest_entry) && (Z_REFCOUNT_P(dest_entry) % 2))) {
+				if (thash && GC_IS_RECURSIVE(thash)) {
 					zend_throw_error(NULL, "Recursion detected");
 					return 0;
 				}
 
-				ZEND_ASSERT(!Z_ISREF_P(dest_entry) || Z_REFCOUNT_P(dest_entry) > 1);
```

Comment:
I got confused this, but it turns out these are leftovers from old refactorings. `Z_ISREF_P(dest_entry)` is never true as we separate it above. So we can remove these.